tedana/decomposition/gift_pca.py

```diff
@@ -0,0 +1,631 @@
+
+    # Using 12 gaussian components from middle, top and bottom gaussian
+    # components to determine the subsampling depth. Final subsampling depth is
+    # determined using median
+    kurtv1 = _kurtn(dataN)
+    kurtv1[EigenValues > np.mean(EigenValues)] = 1000
+    idx_gauss = np.where(
+        ((kurtv1[:, 0] < 0.3) * (kurtv1[:, 0] > 0) * (EigenValues > np.finfo(float).eps)
+         ) == 1)[0]  # DOUBT: make sure np.where is giving us just one tuple
+    idx = np.array(idx_gauss[:]).T
+    dfs = len(
+        np.where(EigenValues > np.finfo(float).eps)[0])  # degrees of freedom
+    minTp = 12
+
+    if (len(idx) >= minTp):
+        middle = int(np.round(len(idx) / 2))
+        idx = np.hstack([idx[0:4], idx[middle - 1:middle + 3], idx[-4:]])
+    else:
+        minTp = np.min([minTp, dfs])
+        idx = np.arange(dfs - minTp, dfs)
+
+    idx = np.unique(idx)
+
+    # Estimate the subsampling depth for effectively i.i.d. samples
+    LGR.info('Estimating the subsampling depth for effective i.i.d samples...')
+    mask_ND = np.reshape(maskvec, (Nx, Ny, Nz), order='F')
+    sub_depth = len(idx)
+    sub_iid_sp = np.zeros((sub_depth, ))
+    for i in range(sub_depth):
+        x_single = np.zeros(Nx * Ny * Nz)
+        x_single[maskvec == 1] = dataN[:, idx[i]]
+        x_single = np.reshape(x_single, (Nx, Ny, Nz), order='F')
+        sub_iid_sp[i] = _est_indp_sp(x_single)[0] + 1
+        if i > 6:
+            tmp_sub_sp = sub_iid_sp[0:i]
+            tmp_sub_median = np.round(np.median(tmp_sub_sp))
+            if np.sum(tmp_sub_sp == tmp_sub_median) > 6:
+                sub_iid_sp = tmp_sub_sp
+                break
+        dim_n = x_single.ndim
+
+    sub_iid_sp_median = int(np.round(np.median(sub_iid_sp)))
+    if np.floor(np.power(np.sum(maskvec) / Nt, 1 / dim_n)) < sub_iid_sp_median:
+        sub_iid_sp_median = int(np.floor(np.power(np.sum(maskvec) / Nt, 1 / dim_n)))
+    N = np.round(np.sum(maskvec) / np.power(sub_iid_sp_median, dim_n))
+
+    if sub_iid_sp_median != 1:
+        mask_s = _subsampling(mask_ND, sub_iid_sp_median)
+        mask_s_1d = np.reshape(mask_s, np.prod(mask_s.shape), order='F')
+        dat = np.zeros((int(np.sum(mask_s_1d)), Nt))
+        LGR.info('Generating subsampled i.i.d. OC data...')
+        for i in range(Nt):
+            x_single = np.zeros((Nx * Ny * Nz, ))
+            x_single[maskvec == 1] = data[:, i]
+            x_single = np.reshape(x_single, (Nx, Ny, Nz), order='F')
+            dat0 = _subsampling(x_single, sub_iid_sp_median)
+            dat0 = np.reshape(dat0, np.prod(dat0.shape), order='F')
+            dat[:, i] = dat0[mask_s_1d == 1]
+
+        # Perform Variance Normalization
+        dat = scaler.fit_transform(dat)
+
+        # (completed)
+        LGR.info('Performing SVD on subsampled i.i.d. OC data...')
+        [V, EigenValues] = _icatb_svd(dat, Nt)
+        LGR.info('SVD done on subsampled i.i.d. OC data')
+        EigenValues = EigenValues[::-1]
+
+    lam = EigenValues
```

Comment:
Why are we re-assigning these here?

Reply:
That was done to follow the MATLAB code. Got rid of it as we are now sure it works.